icu4c/source/i18n/unitsdata.cpp

```diff
@@ -359,32 +380,20 @@ U_I18N_API UnitPreferences::UnitPreferences(UErrorCode &status) {
     ures_getAllItemsWithFallback(unitsBundle.getAlias(), "unitPreferenceData", sink, status);
 }
 
+// TODO: make outPreferences const?
+//
+// TODO: consider replacing `UnitPreference **&outPrefrences` with slice class
+// of some kind.
 void U_I18N_API UnitPreferences::getPreferencesFor(const char *category, const char *usage,
                                                    const char *region,
-                                                   MaybeStackVector<UnitPreference> *outPreferences,
-                                                   UErrorCode &status) {
-    // UnitPreferenceMetadata *m = getMetadata(category, usage, region);
+                                                   const UnitPreference **&outPreferences,
+                                                   int32_t &preferenceCount, UErrorCode &status) const {
     int32_t idx = getPreferenceMetadataIndex(&metadata_, category, usage, region, status);
     if (U_FAILURE(status)) { return; }
     U_ASSERT(idx >= 0); // Failures should have been taken care of by `status`.
-    UnitPreferenceMetadata *m = metadata_[idx];
-    for (int32_t pref = m->prefsOffset; pref < m->prefsOffset + m->prefsCount; pref++) {
-        UnitPreference *p = unitPrefs_[pref];
-        // TODO(review): we're making a full copy of the preferences here.
-        // Considering UnitPreferences instances should simply stick around, we
-        // could also simply return pointers at these instances. What is the
-        // appropriate data structure (array/vector) for variable set of
-        // pointers? MaybeStackVector<UnitPreference*> could probably work, but
-        // ugly as a double-dereference?)
-        UnitPreference *outP = outPreferences->emplaceBack();
-        if (!outP) {
-            status = U_MEMORY_ALLOCATION_ERROR;
-            return;
-        }
-        outP->unit.copyFrom(p->unit, status);
-        outP->geq = p->geq;
-        outP->skeleton.copyFrom(p->skeleton, status);
-    }
+    const UnitPreferenceMetadata *m = metadata_[idx];
+    outPreferences = const_cast<const UnitPreference **>(unitPrefs_.getAlias()) + m->prefsOffset;
```

Comment:
Eew; please find a way to avoid the const_cast.

Reply:
Thanks for answering that question. :-) (The "what about const_cast?" one.) I tried various resolutions for const, it looks like I need to propagate it up many steps, all the way into MaybeStackArray, so I sent this const_cast approach for review first...

Now busy taking another stab at propagating const up the class hierarchy.